File: app/src/main/kotlin/com/absinthe/libchecker/utils/elf/source/ByteBufferDataSource.java
package com.absinthe.libchecker.utils.elf.source;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public class ByteBufferDataSource implements DataSource {
    private final ByteBuffer buffer;
    private ByteOrder order = ByteOrder.LITTLE_ENDIAN;

    public ByteBufferDataSource(ByteBuffer buffer) {
        this.buffer = buffer;
    }

    @Override
    public void read(ByteBuffer dst, long offset) throws IOException {
        buffer.position((int) offset);
        dst.put(buffer);
        dst.flip();
    }

    @Override
    public void read(long offset, byte[] buf, int length) throws IOException {
        buffer.position((int) offset);
        buffer.get(buf, 0, length);
    }

    @Override
    public long readWord(long offset) throws IOException {
        return buffer.getInt((int) offset) & 0xFFFFFFFFL;
    }

    @Override
    public int readHalf(long offset) throws IOException {
        return buffer.getShort((int) offset) & 0xFFFF;
    }

    @Override
    public long readLong(long offset) throws IOException {
        return buffer.getLong((int) offset);
    }

    @Override
    public int readInt(long offset) throws IOException {
        return buffer.getInt((int) offset);
    }

    @Override
    public short readByte(long offset) throws IOException {
        return (short) (buffer.get((int) offset) & 0xFF);
    }

    @Override
    public String readString(long offset) throws IOException {
        StringBuilder sb = new StringBuilder();
        short c;
        while ((c = readByte(offset++)) != 0) {
            sb.append((char) c);
        }
        return sb.toString();
    }

    @Override
    public void position(long offset) throws IOException {
        buffer.position((int) offset);
    }

    @Override
    public ByteOrder getOrder() {
        return order;
    }

    @Override
    public void setOrder(ByteOrder order) {
        this.order = order;
        this.buffer.order(order);
    }

    @Override
    public void close() throws IOException {
        // Nothing to do
    }
}

File: app/src/main/kotlin/com/absinthe/libchecker/utils/elf/source/ByteBufferDataSource.kt
package com.absinthe.libchecker.utils.elf.source

import java.io.IOException
import java.nio.ByteBuffer
import java.nio.ByteOrder

class ByteBufferDataSource(private val buffer: ByteBuffer) : DataSource {
    private var _order: ByteOrder = ByteOrder.LITTLE_ENDIAN

    @Throws(IOException::class)
    override fun read(dst: ByteBuffer, offset: Long) {
        buffer.position(offset.toInt())
        dst.put(buffer)
        dst.flip()
    }

    @Throws(IOException::class)
    override fun read(offset: Long, buf: ByteArray, length: Int) {
        buffer.position(offset.toInt())
        buffer.get(buf, 0, length)
    }

    @Throws(IOException::class)
    override fun readWord(offset: Long): Long {
        return buffer.getInt(offset.toInt()) and 0xFFFFFFFFL
    }

    @Throws(IOException::class)
    override fun readHalf(offset: Long): Int {
        return buffer.getShort(offset.toInt()) and 0xFFFF
    }

    @Throws(IOException::class)
    override fun readLong(offset: Long): Long {
        return buffer.getLong(offset.toInt())
    }

    @Throws(IOException::class)
    override fun readInt(offset: Long): Int {
        return buffer.getInt(offset.toInt())
    }

    @Throws(IOException::class)
    override fun readByte(offset: Long): Short {
        return (buffer.get(offset.toInt()) and 0xFF.toByte()).toShort()
    }

    @Throws(IOException::class)
    override fun readString(offset: Long): String {
        val sb = StringBuilder()
        var currentOffset = offset
        var c: Short
        while (readByte(currentOffset++).also { c = it } != 0.toShort()) {
            sb.append(c.toInt().toChar())
        }
        return sb.toString()
    }

    @Throws(IOException::class)
    override fun position(offset: Long) {
        buffer.position(offset.toInt())
    }

    override val order: ByteOrder
        get() = _order

    override fun setOrder(order: ByteOrder) {
        this._order = order
        this.buffer.order(order)
    }

    @Throws(IOException::class)
    override fun close() {
        // Nothing to do
    }
}